velox/functions/sparksql/tests/HashTest.cpp

```diff
@@ -128,5 +134,98 @@ TEST_F(HashTest, Float) {
   EXPECT_EQ(hash<float>(-limits::infinity()), 427440766);
 }
 
+TEST_F(HashTest, array) {
+  assertEqualVectors(
+      makeFlatVector<int32_t>({2101165938, 42, 1045631400}),
+      hash(makeArrayVector<int64_t>({{1, 2, 3, 4, 5}, {}, {1, 2, 3}})));
+
+  assertEqualVectors(
+      makeFlatVector<int32_t>({-559580957, 1765031574, 42}),
+      hash(makeNullableArrayVector<int32_t>(
+          {{1, std::nullopt}, {std::nullopt, 2}, {std::nullopt}})));
+
+  // Nested array.
+  {
+    auto arrayVector = makeNestedArrayVectorFromJson<int64_t>(
+        {"[[1, null, 2, 3], [4, 5]]",
+         "[[1, null, 2, 3], [6, 7, 8]]",
+         "[[]]",
+         "[[null]]",
+         "[null]"});
+    assertEqualVectors(
+        makeFlatVector<int32_t>({2101165938, -992561130, 42, 42, 42}),
+        hash(arrayVector));
+  }
+
+  // Array of map.
+  {
+    using S = StringView;
+    using P = std::pair<int64_t, std::optional<S>>;
+    std::vector<P> a{P{1, S{"a"}}, P{2, std::nullopt}};
+    std::vector<P> b{P{3, S{"c"}}};
+    std::vector<std::vector<std::vector<P>>> data = {{a, b}};
+    auto arrayVector = makeArrayOfMapVector<int64_t, S>(data);
+    assertEqualVectors(
+        makeFlatVector<int32_t>(std::vector<int32_t>{-718462205}),
+        hash(arrayVector));
+  }
+
+  // Array of row.
+  {
+    std::vector<std::vector<std::optional<std::tuple<int32_t, std::string>>>>
+        data = {
+            {{{1, "red"}}, {{2, "blue"}}, {{3, "green"}}},
+            {{{1, "red"}}, std::nullopt, {{3, "green"}}},
+            {std::nullopt},
+        };
+    auto arrayVector = makeArrayOfRowVector(data, ROW({INTEGER(), VARCHAR()}));
+    assertEqualVectors(
+        makeFlatVector<int32_t>({-1458343314, 551500425, 42}),
+        hash(arrayVector));
+  }
+}
+
+TEST_F(HashTest, map) {
+  auto mapVector = makeMapVector<int64_t, double>(
+      {{{1, 17.0}, {2, 36.0}, {3, 8.0}, {4, 28.0}, {5, 24.0}, {6, 32.0}}});
+  assertEqualVectors(
+      makeFlatVector<int32_t>(std::vector<int32_t>{1263683448}),
+      hash(mapVector));
+
+  auto mapOfArrays = createMapOfArraysVector<int32_t, int32_t>(
+      {{{1, {{1, 2, 3}}}}, {{2, {{4, 5, 6}}}}, {{3, {{7, 8, 9}}}}});
+  assertEqualVectors(
+      makeFlatVector<int32_t>({-1818148947, 529298908, 825098912}),
+      hash(mapOfArrays));
+
+  auto mapWithNullArrays = createMapOfArraysVector<int64_t, int64_t>(
+      {{{1, std::nullopt}}, {{2, {{4, 5, std::nullopt}}}}, {{3, {{}}}}});
+  assertEqualVectors(
+      makeFlatVector<int32_t>({-1712319331, 2060637564, 519220707}),
+      hash(mapWithNullArrays));
+}
+
+TEST_F(HashTest, row) {
+  auto row = makeRowVector({
```

Comment:
@mbasmanova In this example, the first row of this struct is (1, 2), and the second row is (3,  4). According to Spark's implementation, it first compute the first field of the first row, which is value of `hash(1, seed)`, assuming it's `v1`. Then, use `v1` as the seed to compute the second field of the first row `hash(2, v1)`, and then compute the first field of the second row, and so on. 

> First, compute hashes for the first field for all rows. Then, use the results to compute hashes for the second field for all rows. 

This way will break the hash computing order. Because the previous hash result is used as the seed of the next hash computation, breaking the order will change the final result.

Reply:
@marin-ma In this example we have a struct with 2 fields.  To compute hashes in columnar way, we would do the following:

1. Compute hashes for the first field using seed 1. 

```
std::vector<int32_t> h;
h[0] = hash(1, seed)
h[1] = hash(3, seed)  
```

2. Compute hashes or the second field:

```
h[0] = hash(3, h[0])
h[1] = hash(4, h[1])  
```

As long as there are no dependencies between rows, this should work, no?

Reply:
I'm afraid it's not the case. The dependency is from per field and per row. 

here is the logic of Spark:
```
        while (i < map.numElements()) {
          result = hash(keys.get(i, kt), kt, result)
          result = hash(values.get(i, vt), vt, result)
          i += 1
        }
```

h[1] = hash(3, hash(2, hash(1,seed) ) )

in stead of hash (3, seed)

Reply:
@FelixYBW Why do you say this is not the case. Taking this function as an example,

```
  int64_t hashInt32(const int32_t input, uint64_t seed) {
    int64_t hash = seed + PRIME64_5 + 4L;
    hash ^= static_cast<int64_t>((input & 0xFFFFFFFFL) * PRIME64_1);
    hash = bits::rotateLeft64(hash, 23) * PRIME64_2 + PRIME64_3;
    return fmix(hash);
  }
```

looks like `f(input) : = static_cast<int64_t>((input & 0xFFFFFFFFL) * PRIME64_1)` and g is

```'
  int64_t g(f_result, uint64_t seed) {
    int64_t hash = seed + PRIME64_5 + 4L;
    hash ^= f_result;
    hash = bits::rotateLeft64(hash, 23) * PRIME64_2 + PRIME64_3;
    return fmix(hash);
  }
```

Reply:
I see, so you break the hash functions into two parts, f is irrelative to seed so calculate vectorize, and g has dependency on seeds to calculate sequentially.

Hard to say if we can gain performance. If f_result is too large to be hold in cache, there will be one more memory write and load. Also the g loop will be the bottleneck.

We can benchmark it.

Reply:
@mbasmanova
Assume we flatten the fields of complex data type and formed a 1-D vector, then we can have two ways:

f+g:
```
        for (uint32_t i = 0; i < DATA_SIZE; i++)
        {
            f_rst[i] = f(val[i]);
        }

        for (uint32_t i = 0; i < DATA_SIZE; i++)
        {
            seed = g(f_rst[i], seed);
        }
```

fg:
```
        for (uint32_t i = 0; i < DATA_SIZE; i++)
        {
            f_rst = f(val[i]);
            seed = g(f_rst, seed);
        }
```

**conclusion:** 
1. f+g has worse performance,
2. f loop takes 18% time on large data set and takes 8% on small data set. So the gain of vectorize f loop isn't much. g loop is the bottleneck. one reason is f loop is much simpler, another reason is it actually run in parallel in CPU on different pipeline.
3. on cache hit, g loop needs 18 cycles, but if we put f&g together, the multiple in f is actually hidden and the cycles are still 18. Which means no matter how you optimize f loop, vectorize it, there won't be perf gain.

We can definitely optimize the perf further like add software prefetch. but I think we may stop here until we noted it does be a big hotspot in real use case.

Here is the raw data. Depends on the DATA_SIZE which impacted cache misses a lot, here is the result:

  | f+g | fg | f+g | fg
-- | -- | -- | -- | --
data size | 1,000,000,000 | 1,000,000,000 | 1,000,000 | 1,000,000
cycles | 21,184,853,020 | 18,369,158,508 | 16,516,539 | 15,159,275
longest_lat_cache.miss | 238,860,878 | 87,041,651 | 211,535 | 710,252
mem_load_l3_miss_retired.local_dram | 943,210 | 71,831 | 30,008 | 2,041
mem_load_retired.fb_hit | 545,613,986 | 7,150,940 | 48,898,206 | 307,368
mem_load_retired.l1_hit | 1,363,048,935 | 714,015,780 | 95,157,520 | 475,260,671
mem_load_retired.l1_miss | 68,741,636 | 1,356,861 | 6,099,997 | 230,894
mem_load_retired.l2_hit | 39,083,208 | 176,581 | 4,522,438 | 152,547
mem_load_retired.l2_miss | 30,092,393 | 1,339,804 | 1,131,354 | 66,493
mem_load_retired.l3_hit | 89,606 | 47,885 | 1,009,113 | 39,935
mem_load_retired.l3_miss | 29,916,378 | 1,277,610 | 56,605 | 26,299
l2_lines_in.all | 244,983,747 | 87,446,733 | 26,700,625 | 59,113,666
l2_rqsts.miss | 240,175,750 | 87,303,420 | 25,219,720 | 58,880,876
  |   |   |   |  
f loop time | 18% |   | 8% |  
g loop | 80% | 100% | 90% | 100%